modules/atomic-widgets/props-handler/props-handler.php

```diff
@@ -0,0 +1,90 @@
+<?php
+
+namespace Elementor\Modules\AtomicWidgets\PropsHandler;
+
+use Elementor\Modules\AtomicWidgets\Schema\Atomic_Prop;
+use Exception;
+
+if ( ! defined( 'ABSPATH' ) ) {
+	exit; // Exit if accessed directly.
+}
+
+class Props_Handler {
+	/**
+	 * Each transformer can return a value that is also a transformable value,
+	 * which means that it can be transformed again by another transformer.
+	 * this constant defines the maximum depth of transformations.
+	 */
+	const TRANSFORM_DEPTH_LIMIT = 3;
+
+	const CONTEXT_SETTINGS = 'settings';
+	const CONTEXT_STYLE = 'style';
+
+	/**
+	 * @var array<Props_Handler>
+	 */
+	private static array $instances = [];
+
+	private Transformers_Registry $transformers;
+
+	public function __construct( Transformers_Registry $transformers ) {
+		$this->transformers = $transformers;
+	}
+
+	public static function instance( string $context, bool $fresh = false ): self {
+		if ( ! isset( self::$instances[ $context ] ) || $fresh ) {
+			$registry = new Transformers_Registry();
+
+			do_action( "elementor/atomic-widgets/{$context}/transformers", $registry );
+
+			self::$instances[ $context ] = new self( $registry );
+		}
+
+		return self::$instances[ $context ];
+	}
+
+	public function handle( array $props, array $schema ): array {
+		$result = [];
+
+		foreach ( $schema as $prop_name => $prop_type ) {
+			$result[ $prop_name ] = $prop_type instanceof Atomic_Prop
+				? $this->transform( $props[ $prop_name ] ?? $prop_type->get_default() )
+				: null;
+		}
+
+		return $result;
+	}
+
+	private function transform( $value, int $depth = 0 ) {
+		if ( ! $value || ! $this->is_transformable( $value ) ) {
+			return $value;
+		}
+
+		if ( $depth >= self::TRANSFORM_DEPTH_LIMIT ) {
+			return null;
+		}
+
+		$transformer = $this->transformers->get( $value['$$type'] );
+
+		if ( ! ( $transformer instanceof Transformer ) ) {
+			return null;
+		}
+
+		try {
+			return $this->transform(
+				$transformer->transform( $value['value'] ),
+				$depth + 1
+			);
```

Comment:
```suggestion
			$transformed_value = $transformer->transform( $value['value'] );
			
			return $this->transform( $transformed_value, $depth + 1 );
```

I would maybe do this for readability, but not a blocker for me

Reply:
Done